src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_LIST_LENGTH = 10;
+    public static final int ONE_ELEMENT = 1;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_LIST_LENGTH];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeed(ONE_ELEMENT);
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        growIfNeed(ONE_ELEMENT);
+        System.arraycopy(elementData, index, elementData, index + ONE_ELEMENT,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfNeed(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T[] elementDataCopy = elementData;
+        T deletedElement = elementData[index];
+        System.arraycopy(elementData, index + ONE_ELEMENT, elementDataCopy, index,
+                size - ONE_ELEMENT - index);
+        size--;
+        return deletedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        return remove(getElementIndex(element));
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void growIfNeed(int quantity) {
```

Comment:
```suggestion
    public void growIfNeeded(int quantity) {
```
Also pay attention to the access modifier of this and following methods. They shouldn't be public